include/merlin/core_kernels.cuh

```diff
@@ -1448,44 +1448,61 @@ __global__ void remove_kernel(const Table<K, V, M>* __restrict table,
 }
 
 /* Dump with meta. */
+template <class K, class V, class M>
+inline std::tuple<size_t, size_t> dump_kernel_shared_memory_size(
+    const size_t available_shared_memory) {
+  struct KVM {
+    K key;
+    const V* value;
+    M meta;
+  };
+
+  const size_t block_size{
+      std::min(available_shared_memory / 2 / sizeof(KVM), UINT64_C(1024))};
+  MERLIN_CHECK(
+      block_size > 0,
+      "[HierarchicalKV] block_size <= 0, the K-V-M size may be too large!");
+
+  return {block_size * sizeof(KVM), block_size};
+}
+
 template <class K, class V, class M>
 __global__ void dump_kernel(const Table<K, V, M>* __restrict table, K* d_key,
                             V* __restrict d_val, M* __restrict d_meta,
                             const size_t offset, const size_t search_length,
                             size_t* d_dump_counter) {
   extern __shared__ unsigned char s[];
-  const size_t bucket_max_size = table->bucket_max_size;
-  const size_t dim = table->dim;
-  K* smem = (K*)s;
-  K* block_result_key = smem;
-  V* block_result_val = (V*)&(smem[blockDim.x]);
-  M* block_result_meta = (M*)&(block_result_val[blockDim.x * dim]);
+  struct KVM {
```

Comment:
one definition for one thing

Reply:
Yes, I was expecting this comment ;-)

Reply:
Actually, can CUDA use `std::tuple` ? Because `cuda::std::tuple` requires super current CUDA framework.

Reply:
Then we could avoid the definition... ;-(

Reply:
> 
cuda::std::tuple requires `libcxx1.3.0 / CUDA 11.2`, I think it would be OK.

Reply:
I tried to compile with current setup and failed.